### PR TITLE
web-wallet: fix rounding errors in migration amount input

### LINKED ISSUE #2303: web-wallet: Fix rounding display error in migration amount input
#### Summary
When using the `USE MAX` button to set the amount to the maximum available in the account a rounding error occurs as seen below.
This should not happen and the amount introduced in the input field should be the same as the one displayed in the balance field.

image

### PR DESCRIPTION
**Pull request by deuch13**
Resolves #2303

### CODE REVIEW

File: web-wallet/src/lib/dusk/string/cleanNumberString.js [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+ * Returns a valid number in string form with the correct decimal separator according to the locale
+ *
+ * @param {string} amount
+ * @param {string} separator
+ * @returns {string}
+ */
+const cleanANumberString = (amount, separator) => {
+  const regex = new RegExp(`[^\\d${separator}]+`, "g"); // Remove any character that are not digits or the decimal separator
+  const regex2 = new RegExp(`(?<=\\${separator}.*)\\${separator}`, "g"); // Remove all but the first decimal separator
+  const regex3 = new RegExp(`\\${separator}+`, "g"); //Collapse multiple decimal separator into one
```
Comment (ascartabelli) [inline web-wallet/src/lib/dusk/string/cleanNumberString.js:12]: Don't know if it's me, but I'd rather keep the last separator rather than the first (using a look ahead instead of a look behind?).

Having said that: is the third RegExp necessary at all? Shouldn't multiple separators already stripped out by the second one?

File: web-wallet/src/lib/dusk/string/__tests__/cleanNumberString.spec.js
Comment (ascartabelli) [inline web-wallet/src/lib/dusk/string/__tests__/cleanNumberString.spec.js:1]: Would be nice to add some edge case like an empty string, a string containing only non valid chars.

Also you never added invalid chars in the existing tests and you never mixed up separators.

File: web-wallet/src/lib/dusk/number/getDecimalSeparator.js [hunk 2]
```diff
@@ -0,0 +1,14 @@
+/**
+ * @returns {string}
+ */
+
+const getDecimalSeparator = () => {
```
Comment (ascartabelli) [inline web-wallet/src/lib/dusk/number/getDecimalSeparator.js:5]: Wouldn't a `(.1).toLocaleString().slice(1,2)` one-liner be simpler?

File: web-wallet/src/lib/components/ExecuteMigration/ExecuteMigration.svelte [hunk 3]
```diff
@@ -34,7 +37,7 @@
   /** @param {number} id - the chain id of the selected smart contract */
   async function handleMigration(id) {
     const txHash = await migrate(
-      amount.toString(),
+      parseUnits(amount, decimals),
```
Comment (ascartabelli) [inline web-wallet/src/lib/components/ExecuteMigration/ExecuteMigration.svelte:40]: The `parseUnits` function doesn't seem to take into consideration string with decimal separators different than the dot and strings with thousands separators. (see https://github.com/wevm/viem/blob/main/src/utils/unit/parseUnits.ts#L21)
Maybe we should compose it with the clean function you wrote?

File: web-wallet/src/lib/containers/MigrateContract/MigrateContract.svelte [hunk 4]
```diff
@@ -41,12 +46,10 @@
 
   const options = ["ERC-20", "BEP-20"];
 
-  const minAmount = 1n ** 9n;
+  const minAmount = "0.000000000000000001";
```
Comment (ascartabelli) [inline web-wallet/src/lib/containers/MigrateContract/MigrateContract.svelte:49]: Maybe add an explanation comment as this seems like a "magic string" otherwise.

**Review by ascartabelli — COMMENTED** [review]
Left some notes to consider.

File: web-wallet/src/lib/dusk/string/cleanNumberString.js (hunk 1 above)
Comment (deuch13) [reply web-wallet/src/lib/dusk/string/cleanNumberString.js:12]: You're right the third one is overkill and not needed. As for which separator to keep I guess it comes down to preference.

File: web-wallet/src/lib/components/ApproveMigration/ApproveMigration.svelte [hunk 5]
```diff
@@ -3,19 +3,22 @@
...
+  import { isHex, parseUnits } from "viem";
   import { createEventDispatcher } from "svelte";
   import { Button, Icon } from "$lib/dusk/components";
   import { account, wagmiConfig } from "$lib/migration/walletConnection";
   import { allowance, approve } from "$lib/migration/migration";
   import { createDataStore } from "$lib/dusk/svelte-stores";
 
-  /** @type {number} */
+  /** @type {string} */
   export let amount;
```
[the remark this answers is not in the document]
Comment (deuch13) [reply web-wallet/src/lib/components/ApproveMigration/ApproveMigration.svelte:14]: Why would we need to give a reference to the type of the variable in the variable name if we already define the type? Seems like overkill to me.

File: web-wallet/src/lib/components/ExecuteMigration/ExecuteMigration.svelte (hunk 3 above)
Comment (deuch13) [reply web-wallet/src/lib/components/ExecuteMigration/ExecuteMigration.svelte:40]: I haven't considered this, I'll update the code.

File: web-wallet/src/lib/dusk/number/getDecimalSeparator.js (hunk 2 above)
Comment (deuch13) [reply web-wallet/src/lib/dusk/number/getDecimalSeparator.js:5]: Simpler, yes, but I think that using the Intl API might be the correct way.

File: web-wallet/src/lib/components/ApproveMigration/ApproveMigration.svelte (hunk 5 above)
[the remark this answers is not in the document]
Comment (ascartabelli) [reply web-wallet/src/lib/components/ApproveMigration/ApproveMigration.svelte:14]: It's just that the name alone screams "number" to me...

The goal anyway wasn't reference the type in the name, but give it more meaning.

Have you also considered to have the father component do the conversion to BigInt and pass that down to child components?

**Review by ascartabelli — APPROVED** [review]
